File: leetcode/spiral_matrix.swift
import Foundation

class Solution {
  func generateMatrix(_ n: Int) -> [[Int]] {
    var matrix = [[Int]](repeating: [Int](repeating: 0, count: n), count: n)
    var counter = 0
    var i = 0

    // 0, 0: 1
    // 0, 1: 2
    // 0, 2: 3
    // 1, 2: 4
    // 2, 2: 5
    // 2, 1: 6
    // 2, 0: 7
    // 1, 0: 8
    // 1, 1: 9
    while true {
      
    }
    
    for i in 0..<n {
      for j in 0..<n {
        matrix[i][j] = counter
        counter += 1
        let _ = "\(i) \(j) \(matrix)"
      }
    }

    return matrix
  }
}

let result = Solution().generateMatrix(5)
print(result)

File: leetcode/spiral_matrix.py
"""
Spiral Matrix
"""
class Solution(object):
    """
    Solution
    """
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        ROWS, COLS = len(matrix), len(matrix[0])
        visited = [[False] * COLS for _ in matrix]
        DIRS = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        r = c = dir = 0
        result = []
        for _ in range(ROWS * COLS):
            result.append(matrix[r][c])
            visited[r][c] = True
            offs_r, offs_c = DIRS[dir]
            next_r, next_c = r + offs_r, c + offs_c
            if 0 <= next_r < ROWS and 0 <= next_c < COLS and not visited[next_r][next_c]:
                r, c = next_r, next_c
            else:
                dir = (dir + 1) % 4
                offs_r, offs_c = DIRS[dir]
                next_r, next_c = r + offs_r, c + offs_c
                r, c = next_r, next_c
        return result
